include/mbgl/gfx/vertex_attribute.hpp

```diff
@@ -0,0 +1,359 @@
+#pragma once
+
+#include <mbgl/gfx/gfx_types.hpp>
+#include <mbgl/renderer/paint_property_binder.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+namespace mbgl {
+
+using mat2 = std::array<double, 2 * 2>;
+using mat3 = std::array<double, 3 * 3>;
+using mat4 = std::array<double, 4 * 4>;
+
+namespace gfx {
+
+class ShaderProgramBase;
+class VertexVectorBase;
+
+class VertexAttribute {
```

Comment:
A documentation comment should be added to this class.

Some (public) methods including the constructors might also benefit from documentation comments, although they are relatively straightforward.